### PR TITLE
[JENKINS-36432 followup] Switch to SecretBytes rather than home grown secret key

### PR DESCRIPTION
**Pull request by stephenc**
Now that [JENKINS-36432](https://issues.jenkins-ci.org/browse/JENKINS-36432) is available, switch to use the `SecretBytes` API rather than home rolled encryption from a `CryptoConfidentialKey`.

Advantages:

* Makes it possible to programmatically construct file credentials
* Makes it possible to supply credentials in XML form, e.g. 

    ```
        <org.jenkinsci.plugins.plaincredentials.impl.FileCredentialsImpl plugin="plain-credentials@1.3-SNAPSHOT">
          <scope>...</scope>
          <id>...</id>
          <description>...</description>
          <fileName>secret.txt</fileName>
          <secretBytes><!-- Base64 encoded content goes here--></secretBytes>
        </org.jenkinsci.plugins.plaincredentials.impl.FileCredentialsImpl>
```
    will be read in from disk and written down in encrypted form next time the credentials are persisted.
* Uses a salted encryption so that the same content will not result in the same encrypted form... but the salt is maintained for any `SecretBytes` instance so that repeated saves of the same credential instance will not cause unnecessary deltas in the credential store xml file.

@reviewbybees @jenkinsci/code-reviewers @jglick

### CODE REVIEW

**Comment by jglick** [comment]
> Makes it possible to supply credentials in XML form, e.g. […] will be read in from disk

Well this was already true before AFAICT.

File: pom.xml [hunk 1]
```diff
@@ -1,63 +1,95 @@
...
+ ~ IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ ~ FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ ~ AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ ~ LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ ~ OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ ~ THE SOFTWARE.
+ -->
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+  <modelVersion>4.0.0</modelVersion>
```
Comment (jglick) [inline pom.xml]: Why are you reformatting? Please construct a minimal diff.

File: src/main/java/org/jenkinsci/plugins/plaincredentials/impl/FileCredentialsImpl.java [hunk 2]
```diff
@@ -25,47 +25,92 @@
...
+     * @param secretBytes the content of the file.
+     * @throws IOException when things go wrong.
+     * @deprecated use {@link #FileCredentialsImpl(CredentialsScope, String, String, FileItem, String, SecretBytes)} for
+     * stapler or {@link #FileCredentialsImpl(CredentialsScope, String, String, String, SecretBytes)} for programatic
+     * instantiation.
+     */
+    @Deprecated
+    public FileCredentialsImpl(@CheckForNull CredentialsScope scope, @CheckForNull String id,
+                               @CheckForNull String description, @Nonnull FileItem file, @CheckForNull String fileName,
+                               @CheckForNull String secretBytes) throws IOException {
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/plaincredentials/impl/FileCredentialsImpl.java]: Better to keep the original parameter name `data` to avoid confusion.

File: src/main/java/org/jenkinsci/plugins/plaincredentials/impl/FileCredentialsImpl.java [hunk 3]
```diff
@@ -74,34 +119,124 @@
...
+                               @CheckForNull String description, @Nonnull FileItem file, @CheckForNull String fileName,
+                               @CheckForNull SecretBytes secretBytes) throws IOException {
+        super(scope, id, description);
+        String name = file.getName();
+        if (name.length() > 0) {
+            this.fileName = name.replaceFirst("^.+[/\\\\]", "");
+            this.secretBytes = SecretBytes.fromBytes(file.get());
+        } else {
+            this.fileName = fileName;
+            this.secretBytes = secretBytes;
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/plaincredentials/impl/FileCredentialsImpl.java]: Recommended to do some assertions here: in the first clause, `secretBytes` should be null; in the second, should be non-null. (Right?)

File: src/main/java/org/jenkinsci/plugins/plaincredentials/impl/FileCredentialsImpl.java [hunk 4]
```diff
@@ -74,34 +119,124 @@
...
+     */
+    @Extension
+    public static class DescriptorImpl extends BaseStandardCredentialsDescriptor {
 
-        @Override public String getDisplayName() {
+        /**
+         * {@inheritDoc}
+         */
+        @Override
+        public String getDisplayName() {
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/plaincredentials/impl/FileCredentialsImpl.java]: All gratuitous. Please avoid making changes unrelated to the stated purpose of the PR.

File: src/main/java/org/jenkinsci/plugins/plaincredentials/impl/StringCredentialsImpl.java [hunk 5]
```diff
@@ -35,6 +35,8 @@
 
 public final class StringCredentialsImpl extends BaseStandardCredentials implements StringCredentials {
 
+    private static final long serialVersionUID = 4239232115673493707L;
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/plaincredentials/impl/StringCredentialsImpl.java]: Unrelated, better to put in a separate PR.

**Comment by stephenc** [comment]
> Well this was already true before AFAICT.

Nope because the data was the Base64 of the *encoded* bytes... and there was no way to access the encoding key.

With SecretBytes, it will read Base64 of unencoded but will only write the encoded Base64 version

File: src/main/java/org/jenkinsci/plugins/plaincredentials/impl/StringCredentialsImpl.java (hunk 5 above)
Comment (stephenc) [inline src/main/java/org/jenkinsci/plugins/plaincredentials/impl/StringCredentialsImpl.java]: build was failing without it due to findbugs

File: src/main/java/org/jenkinsci/plugins/plaincredentials/impl/FileCredentialsImpl.java (hunk 3 above)
Comment (stephenc) [inline src/main/java/org/jenkinsci/plugins/plaincredentials/impl/FileCredentialsImpl.java]: well if you re-upload the content then secretBytes may be non null when there is a file... so only the non-null check is valid

File: src/main/java/org/jenkinsci/plugins/plaincredentials/impl/FileCredentialsImpl.java (hunk 4 above)
Comment (stephenc) [inline src/main/java/org/jenkinsci/plugins/plaincredentials/impl/FileCredentialsImpl.java]: Once I started adding javadoc I decided it was better to just apply the standard formatting as very little remained

File: src/main/java/org/jenkinsci/plugins/plaincredentials/impl/FileCredentialsImpl.java [hunk 6]
```diff
@@ -25,47 +25,92 @@
...
+     * Our logger.
+     */
     private static final Logger LOGGER = Logger.getLogger(FileCredentialsImpl.class.getName());
+    /**
+     * Standardize serialization (this value is for the 1.2 version of the class).
+     */
+    private static final long serialVersionUID = -7448141713963432962L;
 
-    private final @Nonnull String fileName;
-    private final @Nonnull byte[] data;
```
Comment (oleg-nenashev) [inline src/main/java/org/jenkinsci/plugins/plaincredentials/impl/FileCredentialsImpl.java]: +💯

File: src/main/java/org/jenkinsci/plugins/plaincredentials/impl/FileCredentialsImpl.java [hunk 7]
```diff
@@ -74,34 +119,124 @@
...
+        super(scope, id, description);
+        String name = file.getName();
+        if (name.length() > 0) {
+            this.fileName = name.replaceFirst("^.+[/\\\\]", "");
+            this.secretBytes = SecretBytes.fromBytes(file.get());
+        } else {
+            this.fileName = fileName;
+            this.secretBytes = secretBytes;
+        }
+        if (this.fileName == null || this.fileName.isEmpty()) {
```
Comment (oleg-nenashev) [inline src/main/java/org/jenkinsci/plugins/plaincredentials/impl/FileCredentialsImpl.java]: StringUtils.isBlank() ? Otherwise space-only params may cause some issues even if such file names are actually valid

File: src/main/resources/org/jenkinsci/plugins/plaincredentials/impl/FileCredentialsImpl/credentials.jelly [hunk 8]
```diff
@@ -35,7 +35,7 @@ THE SOFTWARE.
                 <f:textbox field="fileName"/>
             </f:invisibleEntry>
             <f:invisibleEntry>
-                <f:textbox field="data"/>
+                <f:textbox field="secretBytes"/>
```
Comment (oleg-nenashev) [inline src/main/resources/org/jenkinsci/plugins/plaincredentials/impl/FileCredentialsImpl/credentials.jelly]: Causes some confusion. I'd rather keep the original var name

Comment (jglick) [inline src/main/resources/org/jenkinsci/plugins/plaincredentials/impl/FileCredentialsImpl/credentials.jelly]: Perhaps not possible due to `readResolve`, unless the field name deliberately differs from the public getter & constructor parameter names (which is permitted but also confusing).

Comment (stephenc) [inline src/main/resources/org/jenkinsci/plugins/plaincredentials/impl/FileCredentialsImpl/credentials.jelly]: @oleg-nenashev it would cause more confusion to retain the previous field name given that the content of the property has changed.

**Comment by stephenc** [comment]
@jglick I have addressed your request for test cases

File: src/main/java/org/jenkinsci/plugins/plaincredentials/impl/FileCredentialsImpl.java (hunk 7 above)
Comment (stephenc) [inline src/main/java/org/jenkinsci/plugins/plaincredentials/impl/FileCredentialsImpl.java]: This is the logic from before just in the new constructor

File: pom.xml [hunk 9]
```diff
@@ -1,63 +1,95 @@
...
+
+  <scm>
+    <connection>scm:git:git://github.com/jenkinsci/${project.artifactId}-plugin.git</connection>
+    <developerConnection>scm:git:git@github.com:jenkinsci/${project.artifactId}-plugin.git</developerConnection>
+    <url>https://github.com/jenkinsci/${project.artifactId}-plugin</url>
+    <tag>HEAD</tag>
+  </scm>
+
+  <properties>
+    <jenkins.version>1.609</jenkins.version>
```
Comment (jglick) [inline pom.xml]: This seems to have been an actual change.

File: pom.xml [hunk 10]
```diff
@@ -1,63 +1,95 @@
...
+      <id>repo.jenkins-ci.org</id>
+      <url>http://repo.jenkins-ci.org/public/</url>
+    </pluginRepository>
+  </pluginRepositories>
+
+  <dependencies>
+    <dependency>
+      <groupId>org.jenkins-ci.plugins</groupId>
+      <artifactId>credentials</artifactId>
+      <version>2.1.5</version>
```
Comment (jglick) [inline pom.xml]: This too.

File: pom.xml [hunk 11]
```diff
@@ -1,63 +1,95 @@
...
+ ~ THE SOFTWARE.
+ -->
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+  <modelVersion>4.0.0</modelVersion>
+
+  <parent>
+    <groupId>org.jenkins-ci.plugins</groupId>
+    <artifactId>plugin</artifactId>
+    <version>2.15</version>
```
Comment (jglick) [inline pom.xml]: And this.

File: src/main/java/org/jenkinsci/plugins/plaincredentials/impl/StringCredentialsImpl.java (hunk 5 above)
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/plaincredentials/impl/StringCredentialsImpl.java]: Because of the parent POM update?

File: src/test/java/org/jenkinsci/plugins/plaincredentials/SecretBytesTest.java [hunk 12]
```diff
@@ -0,0 +1,190 @@
...
+import org.junit.Test;
+import org.jvnet.hudson.test.JenkinsRule;
+import org.jvnet.hudson.test.recipes.LocalData;
+
+import static hudson.cli.CLICommandInvoker.Matcher.succeededSilently;
+import static org.hamcrest.Matchers.*;
+import static org.junit.Assert.*;
+import static org.junit.Assume.*;
+
+public class SecretBytesTest {
```
Comment (jglick) [inline src/test/java/org/jenkinsci/plugins/plaincredentials/SecretBytesTest.java]: Fine, though it would make sense in `FileCredentialsTest` (which probably should have been `FileCredentialsImplTest`).

File: src/test/java/org/jenkinsci/plugins/plaincredentials/SecretBytesTest.java [hunk 13]
```diff
@@ -0,0 +1,190 @@
...
+
+        // now when we re-save the credentials this should encrypt with the instance's secret key
+        SystemCredentialsProvider.getInstance().save();
+        assertThat(FileUtils.readFileToString(new File(r.jenkins.getRootDir(), "credentials.xml")),
+                not(containsString("VGhpcyBpcyBCYXNlNjQgZW5jb2RlZCBwbGFpbiB0ZXh0Cg==")));
+    }
+
+    /**
+     * Verifies that {@link SecretBytes} will treat a Base64 encoded plain text content as the content to be encrypted
+     * with the instance's secret key which gets applied when the {@link FileCredentialsImpl} is written to disk.
```
Comment (jglick) [inline src/test/java/org/jenkinsci/plugins/plaincredentials/SecretBytesTest.java]: Same Javadoc as `loadUnencrypted` yet presumably this is testing something different?

File: src/test/resources/org/jenkinsci/plugins/plaincredentials/SecretBytesTest/loadUnencrypted/config.xml [hunk 14]
```diff
@@ -0,0 +1,2 @@
+<?xml version='1.0' encoding='UTF-8'?>
+<hudson/>
```
Comment (jglick) [inline src/test/resources/org/jenkinsci/plugins/plaincredentials/SecretBytesTest/loadUnencrypted/config.xml]: An empty root `config.xml` is no longer necessary with current test harness versions.

File: src/test/resources/org/jenkinsci/plugins/plaincredentials/SecretBytesTest/migrateLegacyData/config.xml [hunk 15]
```diff
@@ -0,0 +1,2 @@
+<?xml version='1.0' encoding='UTF-8'?>
+<hudson/>
```
Comment (jglick) [inline src/test/resources/org/jenkinsci/plugins/plaincredentials/SecretBytesTest/migrateLegacyData/config.xml]: ditto

File: src/test/resources/org/jenkinsci/plugins/plaincredentials/SecretBytesTest/migrateLegacyData/secret.key [hunk 16]
```diff
@@ -0,0 +1 @@
+479791e961495d820f337e3afd8fcb70aa36d8c4738093de71d13f7fe6ed6fa0
```
Comment (jglick) [inline src/test/resources/org/jenkinsci/plugins/plaincredentials/SecretBytesTest/migrateLegacyData/secret.key]: I do not think this or `InstanceIdentity.KEY` are actually used, are they?

File: pom.xml (hunk 11 above)
Comment (stephenc) [inline pom.xml]: never said it was just a reformat... I thought I did them in separate commits... but it seems they got squashed somehow (sadpanda)

File: src/test/resources/org/jenkinsci/plugins/plaincredentials/SecretBytesTest/loadUnencrypted/config.xml (hunk 14 above)
Comment (stephenc) [inline src/test/resources/org/jenkinsci/plugins/plaincredentials/SecretBytesTest/loadUnencrypted/config.xml]: well the tests were failing.... perhaps I just need to upgrade the test harness in the `pom.xml` to fix that... but I'll leave that as an exercise for a future PR
